src/framework/components/anim/controller.js

```diff
@@ -0,0 +1,533 @@
+                        conditionsMet = conditionsMet && parameter.value !== condition.value;
+                        break;
+                }
+                if (!conditionsMet)
+                    return conditionsMet;
+            }
+            return conditionsMet;
+        },
+        hasExitTime: function() {
+            return !!this.exitTime;
+        }
+    });
+
+    var AnimController = function (animEvaluator, states, transitions, parameters, activate) {
+        this.animEvaluator = animEvaluator;
+        this.states = states.map(function(state) {
+            return new AnimState(
+                state.name,
+                state.speed
+            );
+        });
+        this.transitions = transitions.map((function(transition) {
+            return new AnimTransition(
+                this,
+                transition.from,
+                transition.to,
+                transition.time,
+                transition.priority,
+                transition.conditions,
+                transition.exitTime,
+                transition.transitionOffset,
+                transition.interruptionSource
+            );
+        }).bind(this));
+        this.parameters = parameters;
+        this.initialParameters = JSON.parse(JSON.stringify(parameters));
+        this.previousStateName = null;
+        this.activeStateName = ANIM_STATE_START;
+        this.playing = false;
+        this.activate = activate;
+        
+        this.currTransitionTime = 1.0;
+        this.totalTransitionTime = 1.0;
+        this.isTransitioning = false;
+        this.transitionInterruptionSource = ANIM_INTERRUPTION_SOURCE_NONE;
+        this.transitionPreviousStates = [];
+
+        this.timeInState = 0;
+        this.timeInStateBefore = 0;
+    };
+
+    Object.assign(AnimController.prototype, {
+        _getState: function(stateName) {
+            for (var i = 0; i < this.states.length; i++) {
+                if (this.states[i].name === stateName) {
+                    return this.states[i];
+                }
+            }
+            return null;
+        },
+
+        _setState: function(stateName, state) {
+            for (var i = 0; i < this.states.length; i++) {
+                if (this.states[i].name === stateName) {
+                    this.states[i] = state;
+                }
+            }
+        },
+
+        _getActiveState: function() {
+            return this._getState(this.activeStateName);
+        },
+
+        _setActiveState: function(stateName) {
+            return this.activeStateName = stateName;
+        },
+
+        _getPreviousState: function() {
+            return this._getState(this.previousStateName);
+        },
+
+        _setPreviousState: function(stateName) {
+            return this.previousStateName = stateName;
+        },
+
+        _findTransitionsFromState: function(stateName) {
+            var transitions = this.transitions.filter((function(transition) {
```

Comment:
you might want to consider keeping a member variable which stores map of from->[list of to transitions].

this will give you fast access to the list of transitions and keep search cost down as the graph size grows.

or you could go further and store the states & transitions in a graph (with a map of state->node). then finding the state should be a fast lookup and you immediately have access to the target via the list of connected transitions.

ideally no operation would require iterating over all states or transitions.

Reply:
👍 Will add this